### DISCUSSION TITLE
Prevent rendering with jupyter notebook

### DISCUSSION THREAD

**Discussion by JosiahParry** [quarto-dev/quarto-cli#11542] [Q&A] [answered]
### Description

I have a qmd document that looks like so: 

````
---
title: Using serde with extendr
---

One of the most widely used rust crates is the **ser**ialization and **de**serialization crate [serde](https://serde.rs/). It enables rust developers to write their custom structs to many different file formats such as json, toml, yaml, csv, and many more as well as read directly from them.

`extendr` provides a `serde` feature that can convert R objects into structs and struct into R objects. 

First, modify your `Cargo.toml` to include the serde feature.

```{toml filename=Cargo.toml}
#[dependenices]
extender-api = { version = "*", features = ["serde"] }
```
````

When I try to render this quarto _always_ tries to use jupyter. I do not want to use python or jupyter to render this document. How can i explicitly tell quarto to use the knitr engine instead? 

This results in an error 

```
Starting python3 kernel...Traceback (most recent call last):
  File "/Applications/quarto/share/jupyter/jupyter.py", line 21, in <module>
    from notebook import notebook_execute, RestartKernel
  File "/Applications/quarto/share/jupyter/notebook.py", line 17, in <module>
    import nbformat
ModuleNotFoundError: No module named 'nbformat'
```

This makes sense because I do not have nbformat installed nor do i want to have it installed

**Comment by JosiahParry** [comment]
I can avoid this by adding an empty R code chunk to the document 

````
```{r include = FALSE}
```
````

but this seems a little bit silly. I do not have an active conda environment, virtualenv, or other such environment. So defaulting to jupyter notebooks seems a bit interesting.

**Comment by mcanouil** [comment, marked as answer]
By default, Quarto tries to guess the compute engine to use.
If there is an R code cell, then `knitr` is used else `Jupyter`.

If you don't want any engine because you don't have any computations, you can use `engine: markdown` (if you have code cells, those won't be evaluated of course).

````qmd
---
title: "Quarto Playground"
format: html
engine: markdown
---

This is a playground for Quarto.

{{< lipsum 1 >}}

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}
````

If you prefer to have `knitr`, you can set it explicitly: `engine: knitr`.

**Reply by mcanouil** [reply]
FYI, the engine setting is not yet supported at project-level unfortunately.
So, if you have multiple files, you need to set it in each.
- https://github.com/quarto-dev/quarto-cli/issues/3157

**Comment by cscheid** [comment]
Adding `engine: knitr` to your document should do the trick.

**Comment by JosiahParry** [comment]
Thanks all! This works for me! The docs, from what I could find, only showed jupyter: and knitr: to specify settings. The engine argument is exactly what i need
